controllers/librarypanel_controller.go

```diff
@@ -0,0 +1,519 @@
+		return ctrl.Result{}, fmt.Errorf("failed to apply to all instances: %v", applyErrors)
+	}
+
+	return ctrl.Result{RequeueAfter: libraryPanel.Spec.ResyncPeriod.Duration}, nil
+}
+
+func (r *GrafanaLibraryPanelReconciler) reconcileWithInstance(ctx context.Context, instance *v1beta1.Grafana, cr *v1beta1.GrafanaLibraryPanel, modelWithHash *libraryPanelModelWithHash) error {
+	if instance.IsInternal() {
+		err := ReconcilePlugins(ctx, r.Client, r.Scheme, instance, cr.Spec.Plugins, fmt.Sprintf("%v-librarypanel", cr.Name))
+		if err != nil {
+			return err
+		}
+	} else if instance.IsExternal() && cr.Spec.Plugins != nil {
+		return fmt.Errorf("external grafana instances don't support plugins, please remove spec.plugins from your library panel cr")
+	}
+
+	grafanaClient, err := client2.NewGeneratedGrafanaClient(ctx, r.Client, instance)
+	if err != nil {
+		return err
+	}
+
+	folderUID, err := getFolderUID(ctx, r.Client, cr)
+	if err != nil {
+		return err
+	}
+
+	if folderUID == "" {
+		folderUID, err = r.GetOrCreateFolder(grafanaClient, cr)
+		if err != nil {
+			return err
+		}
+	}
+
+	uid := content.CustomUIDOrUID(cr, fmt.Sprintf("%s", modelWithHash.Model["uid"]))
+	name := fmt.Sprintf("%s", modelWithHash.Model["name"])
+
+	desired := &libraryPanelToReconcile{
+		Name:          name,
+		FolderUID:     folderUID,
+		Kind:          int64(libraryElementTypePanel),
+		UID:           uid,
+		ModelWithHash: modelWithHash,
+		Version:       0, // this will be filled in by computeOperation in case of update
+	}
+
+	op, err := r.computeOperation(ctx, grafanaClient, cr, desired)
+	if err != nil {
+		return err
+	}
+
+	switch op {
+	case libraryElementOperationNoop: // do nothing
+	case libraryElementOperationCreate:
+		// nolint:errcheck
+		_, err = grafanaClient.LibraryElements.CreateLibraryElement(&models.CreateLibraryElementCommand{
+			FolderUID: desired.FolderUID,
+			Kind:      desired.Kind,
+			Model:     desired.ModelWithHash.Model,
+			Name:      desired.Name,
+			UID:       desired.UID,
+		})
+		if err != nil {
+			return err
+		}
+	case libraryElementOperationUpdate:
+		// nolint:errcheck
+		_, err = grafanaClient.LibraryElements.UpdateLibraryElement(desired.UID, &models.PatchLibraryElementCommand{
+			FolderUID: desired.FolderUID,
+			Kind:      desired.Kind,
+			Model:     desired.ModelWithHash.Model,
+			Name:      desired.Name,
+			Version:   desired.Version,
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	instance.Status.LibraryPanels = instance.Status.LibraryPanels.Add(cr.Namespace, cr.Name, uid)
+	return r.Client.Status().Update(ctx, instance)
+}
+
+// computeOperation looks at the current state of Grafana versus the CR and determines
+// whether a create or update operation should take place (or neither.)
+func (r *GrafanaLibraryPanelReconciler) computeOperation(ctx context.Context, client *genapi.GrafanaHTTPAPI, cr *v1beta1.GrafanaLibraryPanel, desired *libraryPanelToReconcile) (libraryElementOperation, error) {
```

Comment:
I think this could be simpler

UIDs are the only reliable indicator, so matching by name is not useful. I'd prefer a control flow similar to [contact points](https://github.com/grafana/grafana-operator/blob/master/controllers/grafanacontactpoint_controller.go#L157) which only checks if the resource exists and creates it if it doesn't